### PR TITLE
Allow Horovod `teardown()` to complete gracefully if exception thrown in callback setup

### PR DESCRIPTION
**Pull request by speediedan**
## What does this PR do?

Fixes: #11751

`HorovodStrategy.teardown()` may not complete gracefully if an exception is thrown before `HorovodStrategy._exit_stack `is set. Providing a simple fix that verifies `_exit_stack` is set before attempting to use it.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [X] Did you list all the **breaking changes** introduced by this pull request?
- [X] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [] Is this pull request ready for review? (if not, please submit in draft mode)
- [] Check that all items from **Before submitting** are resolved
- [] Make sure the title is self-explanatory and the description concisely explains the PR
- [] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -464,6 +464,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed an issue where Horovod teardown() did not complete gracefully if an exception was thrown during callback setup [#11752](https://github.com/PyTorchLightning/pytorch-lightning/pull/11752)
```
Comment (awaelchli) [inline CHANGELOG.md:467]:
```suggestion
- Fixed an issue where `HorovodStrategy.teardown()` did not complete gracefully if an exception was thrown during callback setup [#11752](https://github.com/PyTorchLightning/pytorch-lightning/pull/11752)
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/strategies/horovod.py [hunk 2]
```diff
@@ -197,8 +197,10 @@ def _filter_named_parameters(model: nn.Module, optimizer: Optimizer) -> List[Tup
 
     def teardown(self) -> None:
         super().teardown()
-        self._exit_stack.__exit__(None, None, None)
-        self._exit_stack = None
+        # teardown may be called before `_exit_stack` is set
+        if self._exit_stack:
+            self._exit_stack.__exit__(None, None, None)
+            self._exit_stack = None
```
Comment (ananthsub) [inline pytorch_lightning/strategies/horovod.py:203]: this is a bug that would've been caught with typechecking.

horovod is currently on the exemption list: https://github.com/PyTorchLightning/pytorch-lightning/blob/cc43d07db1ab77385feff04c01f040c5cad805a9/pyproject.toml#L71

@speediedan - could you try removing this line and seeing if there are other typing errors that are surfaced? this file looks llike it has typehint coverage already

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/strategies/horovod.py (hunk 2 above)
Comment (speediedan) [reply pytorch_lightning/strategies/horovod.py:203]: Great point @ananthsub! I see 18 mypy errors (and 2 notes) in `pytorch_lightning/strategies/horovod.py` with the PL mypy config. I'm not actually using Horovod in my PL extension, but am testing for that strategy's use to warn the user (which is my test that picked up the error). I'll def keep an eye out for these in the future but might not be the right person to make the fixes for this file given I'm not fulling testing Horovod.

**Review by ananthsub — APPROVED** [review]
